docs/architecture/adr-001-initial-QGB-p2p-network.md

```diff
@@ -0,0 +1,195 @@
+# ADR 001: Initial QGB P2P network
+
+## Context
+
+The QGB V2 will rely on a P2P network to exchange attestations signatures. This comes after deciding that using the state will result in state bloat, and using block data will always have an opportunity cost for validators. Thus, we will rely on a P2P network for exchanging attestations' signatures between orchestrators and relayers.
+
+An attestation signature can be identified by the `(nonce,orchestrator,type)` tuple:
+
+- `nonce`: nonce of the attestation
+- `orchestrator`: orchestrator address
+- `type`: data commitment confirm or valset confirm
+
+The issue with using a P2P network is that there is no way of knowing the CID of a signature beforehand. So, relayers will not be able to get the confirms directly from the network.
+
+## Possible solutions
+
+The standard solution is to use Libp2p kadDHT implementation for discovery and routing, then using a Bitswap exchange to publish attestation signatures and get them using CIDs. However, in our case, the CIDs will be unknown until the moment of signature creation making it impossible to rely on Bitswap alone. So, we will need to use another mechanism to know the CIDs or to get them.
+
+### Option 1: DHT + Bitswap
+
+Use the DHT as an indexing engine for the confirms and then getting the confirms using bitswap:
+
+```go
+struct SignatureIndex {
+  CID cid.CID
+  IndexSignature []byte // optional: we can write a validator to validate this signature: <https://github.com/libp2p/go-libp2p-record/blob/f093f9649af5edc2edcb3c262bd2d2a4b022d601/validator.go#L27>
+}
+```
+
+**Producer**: Orchestrator, which will be signing attestations:
+
+- Sign an attestation
+- Create the attestation confirm struct
+- Post it to the blockstore
+- Create a `SignatureIndex`
+- Put the Signature index in the DHT.
+
+```go
+attestationConfirm := AttestationConfirm {
+...
+}
+cid, err := cidlink.DefaultLinkSystem().Store(..., attestationConfirm)
+signatureIndex := SignatureIndex {
+  CID: cid,
+  // other fields
+}
+indexKey := "<nonce,orchestrator_addr,type>"
+dht.PutValue(ctx, indexKey, signatureIndex)
+```
+
+The DHT validator should run `ValidateBasics` on the `SignatureIndex` because if a value with a certain key was posted, it can't be overwritten.
+
+**Consumers**: Relayer, which will be querying attestations:
+
+- Create the `SignatureIndex`. All the fields would be known when querying for a certain signature.
+- Query the signature index from the DHT
+- Query the confirm from bitswap
+
+```go
+indexKey := "<nonce,orchestrator_addr,type>"
+index, err := dht.GetValue(ctx, indexKey)
+confirm, err := exchange.GetBlock(ctx, index.CID)
+```
+
+### Limitations
+
+Using the above approach will give us the following limitations:
+
+- Consumers will have to know when to query the network:
+  - This won't be an issue since we can define a delay to start querying: the time when the attestation request was created in the state machine + a few minutes
```

Comment:
if this isn't an issue, then we might not want to list it as a limitation. Also, don't we have to know when to query for all approaches?